python_modules/dagster/dagster/_daemon/run_coordinator/queued_run_coordinator_daemon.py

```diff
@@ -231,45 +231,51 @@ def _get_runs_to_dequeue(
 
         self._logger.info(
             f"Retrieved %d queued runs, checking limits.{locations_clause}",
-            len(queued_runs),
+            len(queued_run_ids),
         )
 
-        # place in order
-        sorted_runs = self._priority_sort(queued_runs)
         tag_concurrency_limits_counter = TagConcurrencyLimitsCounter(
             tag_concurrency_limits, in_progress_runs
         )
 
         batch: List[DagsterRun] = []
-        for run in sorted_runs:
-            if max_concurrent_runs_enabled and len(batch) >= max_runs_to_launch:
-                break
+        # Paginate through our run ids list so we don't need to hold every run
+        # in memory at once. The maximum number of runs we'll hold in memory is
+        # max_runs_to_launch + page_size.
+        page_size = 25
+        for page in range(0, len(queued_run_ids), page_size):
+            run_ids = queued_run_ids[page : page + page_size]
+            # Create a dict of runs keyed by id so we can maintain FIFO ordering
+            runs_by_id = dict(
+                [(run.run_id, run) for run in instance.get_runs(RunsFilter(run_ids=run_ids))]
+            )
 
-            if tag_concurrency_limits_counter.is_blocked(run):
-                continue
+            for run_id in run_ids:
+                run = runs_by_id.pop(run_id)
 
-            location_name = (
-                run.external_job_origin.location_name if run.external_job_origin else None
-            )
-            if location_name and location_name in paused_location_names:
-                continue
+                if tag_concurrency_limits_counter.is_blocked(run):
+                    continue
 
-            tag_concurrency_limits_counter.update_counters_with_launched_item(run)
-            batch.append(run)
+                location_name = (
+                    run.external_job_origin.location_name if run.external_job_origin else None
+                )
+                if location_name and location_name in paused_location_names:
+                    continue
 
-        return batch
+                tag_concurrency_limits_counter.update_counters_with_launched_item(run)
+                batch.append(run)
 
-    def _get_queued_runs(self, instance: DagsterInstance) -> Sequence[DagsterRun]:
-        queued_runs_filter = RunsFilter(statuses=[DagsterRunStatus.QUEUED])
+            # After each page, re-sort our batch and trim it to our max_runs_to_launch
+            batch = self._priority_sort(batch)
```

Comment:
I want to call attention to the fact that we're now sorting for each page instead of sorting a single time. Otherwise, we need to insert into batch in sort order.